examples/step-90/step-90.cc

```diff
@@ -0,0 +1,1262 @@
+    const NonMatching::MeshClassifier<dim> &mesh_classifier;
+    const DoFHandler<dim>                  &level_set_dof_handler;
+    const VectorType                       &level_set;
+    FEValues<dim>                           level_set_fe_values;
+    NonMatching::FEValues<dim>              non_matching_fe_values;
+  };
+
+  template <int dim>
+  struct CopyData
+  {
+    FullMatrix<double>                   cell_matrix;
+    Vector<double>                       cell_rhs;
+    std::vector<types::global_dof_index> local_dof_indices;
+    unsigned int                         cell_index;
+    double                               cell_L2_error_sqr;
+    double                               cell_H1_error_sqr;
+    double                               cell_stab_error_sqr;
+
+    template <class Iterator>
+    void reinit(const Iterator &cell, const unsigned int dofs_per_cell)
+    {
+      cell_matrix.reinit(dofs_per_cell, dofs_per_cell);
+      cell_rhs.reinit(dofs_per_cell);
+      local_dof_indices.resize(dofs_per_cell);
+      cell->get_dof_indices(local_dof_indices);
+    }
+
+    template <class Iterator>
+    void reinit(const Iterator &cell)
+    {
+      cell_index          = cell->active_cell_index();
+      cell_L2_error_sqr   = 0;
+      cell_H1_error_sqr   = 0;
+      cell_stab_error_sqr = 0;
+    }
+  };
+
+  // @sect3{Normal-gradient stabilization form of TraceFEM}
+  // The following class corresponds to the stabilization form,
+  // its contribution to the global matrix and to the error. More specifically,
+  // the method needs_cell_worker() indicates
+  // whether the bilinear form of the stabilization, unlike the main bilinear
+  // form of Laplace-Beltrami operator, needs the bulk cell quadratures. The
+  // cell worker which is useful in an accumulation by MeshWorkers is provided
+  // by the assemble_cell_worker() method. The remaining method
+  // evaluate_cell_worker() computes the stabilization error for the solution
+  // $u_h$, i.e $s_h(u_h,u_h)$. Also note that the method needs_cell_worker()
+  // indicates that the assembly and the evaluation of the form does require a
+  // bulk cell quadrature. This methodology may be utilized in the MeshWorker.
+  // The stabilization scaling is specified by
+  // $\mathrm{stabilization\_parameter}\cdot
+  // h^\mathrm{stabilization\_exponent}$. For elliptic problems with smooth
+  // solutions we can choose any
+  // $-1\leq \mathrm{stabilization\_exponent} \leq 1$ and
+  // a sufficiently large $\mathrm{stabilization\_parameter}$ that depends of
+  // $\Gamma$.
+  template <int dim>
+  class NormalGradientVolumeStabilization
+  {
+  public:
+    NormalGradientVolumeStabilization(VectorType &solution,
+                                      VectorType &level_set)
+      : solution(solution)
+      , level_set(level_set)
+      , stabilization_parameter(1.0)
+      , stabilization_exponent(-1.0)
+    {}
+
+    VectorType &solution;
+    VectorType &level_set;
```

Comment:
This is slightly awkward design. In the functions below, you need to reference the solution and level set *functions*. Each of these functions is identified by two pieces of information: a DoFHandler and a solution vector. But in your design, you split these two things up: You pass the vectors to the constructor and store references to then, and you get the DoFHandler as part of the interface of the worker functions (via the `cell` argument). A better design is to keep things that reference each other together. That would mean making `solution` and `level_set` part of the argument list of the worker functions that require them, rather than storing them separately. Alternatively, you could also pass the `DoFHandler` to the constructor and store a reference to it as well, but I think that the former is my preferred solution.